### DISCUSSION TITLE
Code template library  dependency

### DISCUSSION THREAD

**Discussion by dfauraz** [nextgenhealthcare/connect#5280] [Feature/Enhancement Ideas]
Hi !

We have a few code template libraries and around 400 channels, many of them using these libraries (version 3.9).
The library references the enabled/disabled channels using their IDs.

This is not logical IMHO : it should be the channel referencing the library.
This is not practical also : every time you have a new channel depending on the library, you must update
the library and redeploy it, even if nothing has changed in the code of the library.

In our CICD (maven based), it causes a lock of headaches.
It is as if you would need to update a library in maven, each time a projects adds it as one of its dependencies.

Of course there is the option to add "all new channels" to the code template.
However, it is nice to have some "scoping" of libraries so this is not really a good option either.

Wouldn't it make sense to have the channel referencing the code template instead ?
Is there a rationale behind the current implementation ?
Are we the only one to have this problem ?

Thanks a lot ! :)

**Comment by mjtrac** [comment]
You can set a channel to be dependent on a code template library from within the channel summary tab.  Click "Set dependencies" and it is the first tab in the resulting popup.  When you establish this dependency in channel code, you will still see it when you are working in the code template area.

I don't think it would be typical to refer to channels by id within a code template, unless the purpose of the code template is to send something to a particular channel.  Even in that case, it might be more reasonable to provide the targeted channel as an argument to the code template.

**Reply by dfauraz** [reply]
Hi Mitch! I think there is some misunderstanding here: it is not about coding and referencing channels in code templates (by id or anything else) but it is about CICD. Typically, you create projects in GIT for the Mirth artifacts (XML files) that work together to build some functionality. The XML artifacts are mostly channels and code templates. If you look at the XML for a code template, you see two XML elements "enabledChannelIds" and "disabledChannelIds" that include the channel ids of the corresponding channels. That's where I mean the code template references the channels by id ! A logical project organization would put the code template library in a separate GIT project, referenced by the channel projects. And that's where it is cumbersome with the current implementation : each time you add a dependency in your channel, the library has to be updated as well (version increment. rebuild, retest of all dependencies ...). I hope this helps

**Reply by dfauraz** [reply]
And to make it clear: my suggestion is just to remove the XML elements "enabledChannelIds" and "disabledChannelIds" in the code templates, and replace them with XML elements "enabledLibraryIds" and "disabledLibraryIds" in the channels.

**Reply by dfauraz** [reply]
Another minimal option would be to have the option "all channels" in code template libraries.
Note that "new channels" (which already exists) is not exactly the same: existing channels are not included.

**Reply by dfauraz** [reply]
Note: resources work that way already

**Reply by mjtrac** [reply]
Thanks for the clarification.  That certainly sounds like a reasonable suggestion to me, but I'm not a developer or in product management.
